### PR TITLE
[faqs] Can use full_package_mode over my dependencies

### PR DESCRIPTION
**Pull request by uilianries**
Spoiler: No.

Using `full_package_mode` will result on **missing package** errors on CCI, so it's not recommended. When really needed, I would recommend by using `shared=True` instead and maybe, not allowing static link. 

/cc @SpaceIm 

Related to https://github.com/conan-io/conan-center-index/pull/15756#discussion_r1106041531

---

- [ ] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [ ] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: docs/faqs.md [hunk 1]
```diff
@@ -438,3 +438,23 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+otherwise, it may result on unexpected behavior, or even runtime errors.
+
+A very kwnon project is [glib](https://conan.io/center/glib), which requires the very same configuration to prevent multiple instances when using static link. As solution, we could consume glib on full package id mode, like:
+
+```python
+def package_id(self):
+    self.info.requires["glib"].full_package_mode()
+```
+
+Perfect solution on consumer side, but there is a painful side-effect: CCI will not re-generate all involved packages for any change in the dependencies graph which glib is associated, which means, users will start to see **MISSING_PACKAGES** error during their pull requests.
```
Comment (prince-chrismc) [inline docs/faqs.md:455]: I am concerned the affect 2.0 will have since this is what the embedded mode will do by default

**Comment by SpaceIm** [comment]
There are 18 recipes with `full_package_mode()` currently:

```
grep "full_package_mode()" */*/conanfile.py | awk '{print $1}' | sort -u
aravis/all/conanfile.py:
at-spi2-atk/all/conanfile.py:
at-spi2-core/all/conanfile.py:
at-spi2-core/new/conanfile.py:
atk/all/conanfile.py:
cairo/all/conanfile.py:
cairo/meson/conanfile.py:
cairomm/all/conanfile.py:
gdk-pixbuf/all/conanfile.py:
glibmm/all/conanfile.py:
googleapis/all/conanfile.py:
graphene/all/conanfile.py:
grpc-proto/all/conanfile.py:
grpc/all/conanfile.py:
gst-plugins-base/all/conanfile.py:
gstreamer/all/conanfile.py:
harfbuzz/all/conanfile.py:
pango/all/conanfile.py:
```

There are opened PRs removing them:

aravis: https://github.com/conan-io/conan-center-index/pull/15742
atk: https://github.com/conan-io/conan-center-index/pull/16016
gdk-pixbuf: https://github.com/conan-io/conan-center-index/pull/16012
graphene: https://github.com/conan-io/conan-center-index/pull/15947

File: docs/faqs.md (hunk 1 above)
Comment (uilianries) [reply docs/faqs.md:455]: A sheperds topic then

File: docs/faqs.md [hunk 2]
```diff
@@ -438,3 +438,23 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+As workaround, it would be necessary updating all recipes involved, by opening new PRs, then it should generate new packages, but it takes many days and still is a fragile process.
+
+To have more context about it, please, visit the issues #11684 and #11022
+
+In summary, we do not recommend `full_package_mode` usage on CCI. Instead, prefer using `shared=True` by default, when is really needed.
```
Comment (SSE4) [inline docs/faqs.md:460]:
```suggestion
In summary, we do recommend `full_package_mode` usage on CCI. Instead, prefer using `shared=True` by default, when is really needed.
```
it should be recommended to use, as it's an official documented method

File: docs/faqs.md [hunk 3]
```diff
@@ -438,3 +438,23 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+
+A very kwnon project is [glib](https://conan.io/center/glib), which requires the very same configuration to prevent multiple instances when using static link. As solution, we could consume glib on full package id mode, like:
+
+```python
+def package_id(self):
+    self.info.requires["glib"].full_package_mode()
+```
+
+Perfect solution on consumer side, but there is a painful side-effect: CCI will not re-generate all involved packages for any change in the dependencies graph which glib is associated, which means, users will start to see **MISSING_PACKAGES** error during their pull requests.
+As workaround, it would be necessary updating all recipes involved, by opening new PRs, then it should generate new packages, but it takes many days and still is a fragile process.
```
Comment (SSE4) [inline docs/faqs.md:456]:
```suggestion
As workaround, it would be necessary updating all recipes involved, by opening new PRs, then it should generate new packages, but it takes many days and still is a fragile process.
```
it's not a workaround, it's just a necessarily trade-off of `full_package_mode`. you use it and accept consequences, or you don't use it.

File: docs/faqs.md (hunk 1 above)
Comment (SSE4) [reply docs/faqs.md:455]: I suppose in v2 embed mode should be always active, for C++ and C projects.
as there could be some inline functions in headers, or just macro definitions like:
```
#define SUM(A, B) ((A) + (B))
```
the code from the header is always embedded into the consumer (regardless if it's shared or static library), therefore, if that header changes, all the consumers have to be rebuilt. otherwise, an incompatibility might be encountered (different components use different versions of macro, and such a macro, may define something important for API, like a structure alignment computation).

File: docs/faqs.md [hunk 4]
```diff
@@ -438,3 +438,23 @@ Generally no, these sorts of options can most likely be set from a profile or do
 and would otherwise dynamically embed this into the CMake config files or generated pkg-config files then it should be allowed.
 
 Doing so requires [deleting the option from the `package_id`](adding_packages/conanfile_attributes.md#removing-from-package_id).
+
+## Can I use full_package_mode for a requirement in my recipe?
+
+For some unusual projects, they need to be aligned and being used as requirement, using the very same version, options and settings.
```
Comment (SSE4) [inline docs/faqs.md:444]:
```suggestion
For some usual projects, they need to be aligned and being used as requirement, using the very same version, options and settings.
```

File: docs/faqs.md (hunk 1 above)
Comment (SSE4) [inline docs/faqs.md:455]: that's something to explore for a future, as with embed mode in v2 problem will appear way more often.
possible solution:
- a new tapaholes-like background job to re-generate downstream graph
- allow `--build missing` or `--build xxx` or similar in certain PRs
- allow to build non-standard configurations (e.g. via special keywords in PR title or first message), like overriding default options, versions of dependencies and so on
- expose an interface to build missing (or non-standard) packages on demand (via GH comment, or dedicated web-page)

**Comment by SSE4** [comment]
I think most of GLib and GStreamer ecosystem (plus all the other GTK stuff, and whatever else is based on top of GLib) will suffer from the same problem.

File: docs/faqs.md (hunk 1 above)
Comment (prince-chrismc) [reply docs/faqs.md:455]: It's the new default depending on package type :)

Comment (SSE4) [reply docs/faqs.md:455]: it should be regardless of `package_type`. perhaps, the only exception is header-only depending on header-only.

**Comment by uilianries** [comment]
@SpaceIm @SSE4 I talked with @prince-chrismc and the team. Indeed using full or any package id mode will result in missing packages, so the idea is avoiding it. Of course, anyone is able and allowed to write why they need it. 

For Conan 2.x the situation is like Konstantin said, we will need to re-generate those packages in the dependencies tree, with TapaHoles CI job or another thing, because the package mode will not be the same.

But for those cases where mixing shared with static was a problem, and package id mode was a solution, on Conan 2.x we will distribute shared with shared and static with static by default. 

In summary, we can remove from that list pointed by SpaceIM.

File: docs/faqs.md [hunk 5]
```diff
@@ -438,3 +438,27 @@ Generally no, these sorts of options can most likely be set from a profile or do
 and would otherwise dynamically embed this into the CMake config files or generated pkg-config files then it should be allowed.
 
 Doing so requires [deleting the option from the `package_id`](adding_packages/conanfile_attributes.md#removing-from-package_id).
+
+## Can I use full_package_mode for a requirement in my recipe?
+
+For some not regular projects, they may need to be aligned when being used as a requirement, using the very same version, options, and settings and maybe not mixing shared with static linkage.
```
Comment (SSE4) [inline docs/faqs.md:444]:
```suggestion
For some regular projects, they may need to be aligned when being used as a requirement, using the very same version, options, and settings and maybe not mixing shared with static linkage.
```

File: docs/faqs.md [hunk 6]
```diff
@@ -438,3 +438,27 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+
+A very known project is GLib, which requires the very same configuration to prevent multiple instances when using static linkage.
+As a solution, we could consume glib on full package id mode, like:
+
+```python
+def package_id(self):
+    self.info.requires["glib"].full_package_mode()
+```
+
+Perfect solution on the consumer side, but there is a painful side-effect: CCI will not re-generate all involved packages for any change in the dependencies graph with which glib is associated, which means, users will start to see **MISSING_PACKAGES** error during their pull requests.
```
Comment (SSE4) [inline docs/faqs.md:456]:
```suggestion
Perfect solution on the consumer side, but there is a side-effect: CCI will not re-generate all involved packages for any change in the dependencies graph with which glib is associated, which means, users will start to see **MISSING_PACKAGES** error during their pull requests.
```

File: docs/faqs.md [hunk 7]
```diff
@@ -438,3 +438,27 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+As a trade-off, it would be necessary to update all recipes involved, by opening new PRs,
+then it should generate new packages, but it takes many days and still is a process that is not supported by CCI internally.
+
+To have more context about it, please, visit issues #11684 and #11022
+
+In summary, we do not recommend `full_package_mode` or any other custom package id mode for requirements on CCI, it will break other PRs soon or later.
```
Comment (SSE4) [inline docs/faqs.md:462]:
```suggestion
In summary, we recommend `full_package_mode` or any other custom package id mode for requirements on CCI, it will break other PRs soon or later.
```

File: docs/faqs.md [hunk 8]
```diff
@@ -438,3 +438,27 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
+Instead, prefer using `shared=True` by default, when needed.
```
Comment (SSE4) [inline docs/faqs.md:463]:
```suggestion
You may also try using `shared=True` by default, when needed.
```

File: docs/faqs.md [hunk 9]
```diff
@@ -438,3 +438,27 @@ Generally no, these sorts of options can most likely be set from a profile or do
...
 Doing so requires [deleting the option from the `package_id`](adding_packages/conanfile_attributes.md#removing-from-package_id).
+
+## Can I use full_package_mode for a requirement in my recipe?
+
+For some not regular projects, they may need to be aligned when being used as a requirement, using the very same version, options, and settings and maybe not mixing shared with static linkage.
+Those projects usually break between patch versions and are very sensitive, so we can not use different versions through Conan graph dependencies,
+otherwise, it may result in unexpected behavior or even runtime errors.
+
+A very known project is GLib, which requires the very same configuration to prevent multiple instances when using static linkage.
+As a solution, we could consume glib on full package id mode, like:
```
Comment (SSE4) [inline docs/faqs.md:449]:
```suggestion
As a solution, we could consume GLib on full package id mode, like:
```

File: docs/faqs.md (hunk 5 above)
Comment (SSE4) [reply docs/faqs.md:444]: I wouldn't say GLib is irregular

File: docs/faqs.md (hunk 6 above)
Comment (SSE4) [reply docs/faqs.md:456]: that's too personal and subjective characteristics, I would just operate on raw facts here and avoid emotional descriptions

File: docs/faqs.md (hunk 7 above)
Comment (SSE4) [reply docs/faqs.md:462]: it's fine to recommend if user accepts the consequences described above

Comment (uilianries) [reply docs/faqs.md:462]: It's not fine. The recommendation is NOT using any package ID mode in CCI.

File: docs/faqs.md (hunk 5 above)
Comment (uilianries) [reply docs/faqs.md:444]: It's irregular in terms of consumption. We don't have this problem with most of projects in CCI.

Comment (SSE4) [inline docs/faqs.md:444]:
```suggestion
For some irregular projects, they may need to be aligned when being used as a requirement, using the very same version, options, and settings and maybe not mixing shared with static linkage.
```

File: docs/faqs.md (hunk 7 above)
Comment (SSE4) [reply docs/faqs.md:462]: that's not ideal. we can explain pros and cons of each approach, and let people choose, recommending to use package ID mode if they are agree to accept these consequences of broken PR. it's like "okay, I've been warned, but I acknowledge and take all the risks as my responsibility".

Comment (uilianries) [reply docs/faqs.md:462]: No because you are breaking all other users instead, including CCI CI with missing package, which is worst. The mention is clear about for requirements on CCI, if an user wants to use `full_transitive_mode` locally, fine, but on CCI it's not the idea.

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
In the near future we will need to review the docs and update them for 2.0 and re can remove of put this in a "1.x" section.

The new default package_id should do what this workaround was trying to achieve

**Comment by uilianries** [comment]
> The new default package_id should do what this workaround was trying to achieve

Totally true! 😸

**Review by jwillikers — APPROVED** [review]
